File: bin/photo_csv.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import os
import sys
import json
import re
import csv
import codecs
import unicodedata

json_re = re.compile('\.json$')
files = []
# files.append(so.path.join(os.path.dirname(__file__), '..', 'POP_0001.json'))
data_dir =  os.path.join(os.path.dirname(__file__), '..', 'data')
rows = []

sys.stdout = codecs.getwriter('utf-8')(sys.stdout)
sys.stderr = codecs.getwriter('utf-8')(sys.stderr)

class Book:
    NORMAL_RE  = re.compile('\W')
    FOLIO_RE   = re.compile('^Folio\s*', re.IGNORECASE)

    names = ('call',
             'bibid',
             'title',
             'author',
             'conference',
             'pub',
             'date',
             'provenance',
             'printplace')

    def __init__(self, line):
        self._line = line.rstrip('\r\n')
        self._parts = self._line.split('\t')
        self._props = {}
        # call, bibid, title, author, conference, pub, date, provenance, printplace
        for i in xrange(len(self.names)):
            if len(self._parts) > i:
                self._props[self.names[i]] = self._parts[i]

    @property
    def normal_callno(self):
        s = self._props['call']
        if self.FOLIO_RE.search(s):
            s = self.FOLIO_RE.sub('', s)
            s += 'folio'
        return self.NORMAL_RE.sub('', s.lower())

class Books:

    def __init__(self, tsv):
        self._tsv = tsv
        self._call_to_bibid = {}
        self._books = []
        self._callno_map = {}
        self._index = 0
        self._load_tsv()

    def _load_tsv(self):
        for line in codecs.open(self._tsv, 'r', 'utf-8'):
            try:
                book = Book(line)
                self._books.append(book)
                self._callno_map[book.normal_callno] = book
            except ValueError as ex:
                print "Bad line: %s" % (line,)
                raise ex

    def __iter__(self):
        return self

    def find(self, normal_callno):
        return self._callno_map.get(normal_callno, None)

    def next(self):
        if self._index >= len(self._books):
            raise StopIteration
        self._index += 1
        return self._books[self._index - 1]

class Categories:
    NOT_FOUND = ('NOT_FOUND', 'NOT_FOUND')
    DEFAULT_CAT = 'TBD'

    def __init__(self, csv):
        "Create a new list of categories"
        self.csv = csv
        self.cat_set = set()
        self.tags = {}
        self.load_csv()

    def load_csv(self):
        for line in codecs.open(self.csv, 'r', 'utf-8'):
            try:
                # Ugly	Clean	Mitch's Category	Laura's Category	Model	Field	Basis
                # zaravoni	Zaravoni	Person	Person	Copy	Author	Transcription | Secondary | Imagery Description | Heraldry Description
                ugly, raw, mitch, laura, model, field = line.rstrip('\n').split('\t')
                # ugly, raw, model, field = row[:]
                cat = (model.strip().lower(),field.strip().lower())
                self.tags[raw] = cat
                self.cat_set.add(cat)
            except:
                print "Problem: %r" % line
                raise
        self.cat_set.add(Categories.NOT_FOUND)
        self.cats = list(self.cat_set)
        self.cats.sort()

class Photo:
    def __init__(self,record):
        "Create a photo thingy."
        self.photo_dict = record
        self._tag_map = {}

    @property
    def photo(self):
        return self.photo_dict['photo']

    def _stringify(self, s):
        return str(s.encode('utf-8')).strip()

    @property
    def pid(self):
        # return self._stringify(self.photo['id'])
        return self.photo['id']

    @property
    def title(self):
        # print 'title: %s' % (self.photo['title'], )
        # return self._stringify(self.photo['title'])
        return self.photo['title']

    @property
    def url(self):
        # return self._stringify(self.photo['urls']['url'][0]['text'])
        return self.photo['urls']['url'][0]['text']

    @property
    def tags(self):
        return self.photo['tags']['tag']

    def all_tag_vals(self, key):
        return [ x[key] for x in self.tags]

    @property
    def raw_tags(self):
        return self.all_tag_vals('raw')

    @property
    def text_tags(self):
        return self.all_tag_vals('text')

    def map_tags(self, categories):
        self._tag_map[Categories.NOT_FOUND] = []
        for cat in categories.cats:
            self._tag_map[cat] = []

        for tag in self.raw_tags:
            tag = unicodedata.normalize('NFC', tag)
            if tag in categories.tags:
                cat = categories.tags[tag]
            else:
                cat = Categories.NOT_FOUND
                sys.stderr.write("Tag not found: %s; photo_id: %s\n" % (tag, self.pid))
            self._tag_map[cat].append(tag)

    def mapped_tags(self, categories):
        if len(self._tag_map) == 0:
            self.map_tags(categories)
        return self._tag_map

    def call_nos(self):
        if len(self._tag_map) > 0 and self._tag_map.get((u'copy', u'call number'), None):
            return self._tag_map[(u'copy',u'call number')]
        else:
            return []

cats = Categories(sys.argv[1])
print cats.cats

books = Books(os.path.join(data_dir, 'call_bidid_map.tsv'))

# get the Photo JSON files
for fname in os.listdir(os.path.join(data_dir, 'NFC')):
    if re.search(json_re, fname):
        files.append(fname)

photos = None


head = [ 'ID', 'Title', 'URL', 'BibID' ]
for cat in cats.cats:
    head.append(':'.join(cat))

for fname in files:
    records = json.load(codecs.open(os.path.join(data_dir,'NFC',fname), 'r', 'utf-8'))
    for record in records:
        photo = Photo(record)
        row = [ photo.pid, photo.title, photo.url ]
        tags = photo.mapped_tags(cats)
        bibids = []
        for call in photo.call_nos():
            book = books.find(call)
            if book:
                bibids.append(book._props['bibid'])
        row.append('|'.join(bibids))
        for cat in cats.cats:
            row.append('|'.join(tags[cat]))
        rows.append(row)

outfile = codecs.open("photos_mapped.csv", "w+", 'utf-8')
outfile.write(u'\t'.join(head))
outfile.write('\n')
for row in rows:
    outfile.write(u'\t'.join(row))
    outfile.write('\n')

outfile.close()
print "Wrote \"%s\"" % outfile.name

File: bin/photo_csv.rb
#!/usr/bin/env ruby

require 'json'
require 'set'
require 'pp'
require 'axlsx'

data_dir = File.expand_path '../../data', __FILE__
outfile = 'photos.xlsx'

class Book
  attr_accessor :line, :parts, :props

  NORMAL_RE  = /\W/
  FOLIO_RE   = /^Folio\s*/i

  NAMES = %w(
    call bibid title author conference pub date provenance printplace
  ).map(&:to_sym)

  class << self
    def normalize s
      s and s.strip.gsub(/\W/,'').downcase
    end
  end

  def initialize line
    @line = line.chomp
    @parts = line.split(/\t/)
    @props = {}
    # call, bibid, title, author, conference, pub, date, provenance, printplace
    NAMES.each_with_index do |name,index|
      @props[name] = @parts[index]
    end
  end

  def callno
    if props[:call] =~ FOLIO_RE
      "#{$'} folio"
    else
      props[:call]
    end
  end

  def bibid
    props[:bibid]
  end

  def normal_callno
    Book.normalize callno
  end
end

class Books
  attr_accessor :tsv_file

  def initialize tsv_file
    @tsv_file = tsv_file
    @call_to_bibid = {}
    @books = []
    @callno_map = {}
    @index = 0
    load_tsv
  end

  def load_tsv
    IO.readlines(tsv_file).drop(1).each do |line|
      book = Book.new line
      @books << book
      @callno_map[book.normal_callno] = book
    end
  end

  def each
    return enum_for(:each) unless block_given?

    @books.each do |book|
      yield book
    end
  end

  def find callno
    @callno_map[Book.normalize callno]
  end
end

class Categories
  attr_accessor :csv, :cat_set, :tags, :cats
  NOT_FOUND = %w( NOT_FOUND NOT_FOUND )

  def initialize csv
    @csv = csv
    @cat_set = Set.new
    @tags = {}
    @cats = []
    load_csv
  end

  def load_csv
    IO.foreach csv do |line|
      ugly, raw, mitch, laura, model, field = line.chomp.split "\t"
      cat = [ model.strip.downcase, field.strip.downcase ]
      tags[raw.strip.downcase] = cat
      cat_set.add cat
      cat_set.add(cat)
    end
    @cats = cat_set.to_a
    @cats.sort!
  end
end # class Categories

class Photo
  attr_accessor :record, :tag_hash

  def initialize record, categories
    @record = record
    @tag_hash = {}
    map_tags categories
end

  def photo
    record['photo']
  end

  def photo_id
    photo['id']
  end

  def title
    photo['title']
  end

  def url
    photo['urls']['url'].first['text']
  end

  def tags
    photo['tags']['tag']
  end

  def all_tag_vals key
    tags.map { |t| t[key] }
  end

  def raw_tags
    all_tag_vals 'raw'
  end

  def text_tags
    all_tag_vals 'text'
  end

  def map_tags categories
    (@tag_map ||= {})[Categories::NOT_FOUND] = []
    raw_tags.each do |tag|
      cat = categories.tags[tag.strip.downcase] || Categories::NOT_FOUND
      if cat == Categories::NOT_FOUND
        $stderr.puts "Tag not found: #{tag}; photo_id: #{photo_id}"
      end
      (@tag_map[cat] ||= []) << tag
    end
  end

  def mapped_tags
    @tag_map || {}
  end

  def call_nos
    mapped_tags[['copy','call number']] || []
  end

  def tags_for category
    mapped_tags[category] || []
  end
end

cats = Categories.new ARGV.shift

books = Books.new File.join data_dir, 'call_bidid_map.tsv'

head = %w{ ID Title URL BibID }
cats.cats.reduce(head) { |a,c| a << c.join(':'); a }

p = Axlsx::Package.new
wb = p.workbook
wb.add_worksheet name: 'photos' do |wksh|
  wksh.add_row head
  Dir[File.join data_dir, 'NFC', '*.json'].each do |file|
    records = JSON.load(File.open(file))
    records.each_with_index do |record,index|
      photo = Photo.new record, cats
      row = [ photo.photo_id, photo.title, photo.url ]
      # tags = photo.mapped_tags
      bibids = photo.call_nos.flat_map { |call|
        book = books.find call
        book ? [ book.bibid ] : []
      }
      row << bibids.join('|')
      cats.cats.each do |cat|
        row << photo.tags_for(cat).join('|')
      end
      row = row.map { |x| x == '' ? nil : x }
      wksh.add_row row
    end

    wksh.column_widths(*Array.new(head.size, 20))
  end
end

p.serialize outfile
puts "Wrote '#{outfile}'"
